kernel/kernel-api/src/main/java/io/delta/kernel/internal/ScanImpl.java

```diff
@@ -190,6 +206,56 @@ public void close() throws IOException {
         };
     }
 
+    private Optional<Predicate> getDataSkippingPredicate() {
+        Optional<Predicate> dataPredicate = getDataFilters();
+        if (!dataPredicate.isPresent()) { // no data filter
+            return Optional.empty();
+        }
+        return DataSkippingUtils.constructDataFilters(dataPredicate.get(), metadata.getSchema());
+    }
+
+    private CloseableIterator<FilteredColumnarBatch> applyDataSkipping(
+        TableClient tableClient,
+        CloseableIterator<FilteredColumnarBatch> scanFileIter,
+        Predicate dataFilter) {
+        // Get the stats schema
+        // TODO prune stats schema according to the data filter
+        StructType statsSchema = DataSkippingUtils.getStatsSchema(metadata.getSchema());
+
+        // Skipping happens in two steps:
+        // 1. The predicate produces false for any file whose stats prove we can safely skip it. A
+        //    value of true means the stats say we must keep the file, and null means we could not
+        //    determine whether the file is safe to skip, because its stats were missing/null.
+        // 2. The coalesce(skip, true) converts null (= keep) to true
+        Predicate filterToEval = new Predicate(
+            "=",
+            new ScalarExpression(
+                "COALESCE",
+                Arrays.asList(dataFilter, Literal.ofBoolean(true))),
+            AlwaysTrue.ALWAYS_TRUE);
```

Comment:
@vkorukanti this code `AlwaysTrue.ALWAYS_TRUE` really looks odd. Can we replace this with `Literal.ALWAYS_TRUE` or `Literal.TRUE`?

Reply:
`AlwaysTrue` is a predicate and `Literal.TRUE` is not. Not sure if we can mix predicates and literals together.